src/serde/de_tree.rs

```diff
@@ -0,0 +1,327 @@
+use std::convert::TryInto;
+use std::io::{Error, Read, Result, Write};
+
+use sha2::Digest;
+
+use crate::sha2::Sha256;
+
+use super::parse_atom::decode_size_with_offset;
+use super::utils::{copy_exactly, skip_bytes};
+
+const MAX_SINGLE_BYTE: u8 = 0x7f;
+const CONS_BOX_MARKER: u8 = 0xff;
+
+struct ShaWrapper(Sha256);
+
+impl Write for ShaWrapper {
+    fn write(&mut self, blob: &[u8]) -> std::result::Result<usize, Error> {
+        self.0.update(blob);
+        Ok(blob.len())
+    }
+    fn flush(&mut self) -> std::result::Result<(), Error> {
+        Ok(())
+    }
+}
+
+/// This data structure is used with `parse_triples`, which returns a triple of
+/// integer values for each clvm object in a tree.
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum ParsedTriple {
+    Atom {
+        start: u64,
+        end: u64,
+        atom_offset: u32,
+    },
+    Pair {
+        start: u64,
+        end: u64,
+        right_index: u32,
+    },
+}
+
+enum ParseOpRef {
+    ParseObj,
+    SaveCursor(usize),
+    SaveIndex(usize),
```

Comment:
I don't have a strong opinion, but `Patch` or `Fill` (instead of `Save`) might be slightly more intuitive, given that we actually go back and update entries we've created earlier.

Reply:
Changed.